### PR TITLE
rpc: free buffer after client disconnect

### PR DESCRIPTION
**Pull request by chraac**
In PR #6829, @rgerganov add support to rpc backend, after using it for several days, I have noticed an issue:
- When the client closes the connection, the server does not free the memory it has allocated.

Upon investigating the source code, I discovered that instead of releasing the memory, we simply exit the inner loop and immediately wait for a new connection ([ggml-rpc.cpp#L1027](https://github.com/ggerganov/llama.cpp/blob/f5bf761747988ee1832766f7d1433739aff810da/ggml-rpc.cpp#L1027)).

So here I create this PR, which monitor the `ALLOC_BUFFER` and `FREE_BUFFER` command, maintaining a list of allocated buffers, then free the remind buffer after client disconnect.

### CODE REVIEW

File: ggml-rpc.cpp [hunk 1]
```diff
@@ -934,7 +938,7 @@ static void rpc_serve_client(ggml_backend_t backend, sockfd_t sockfd, size_t fre
         }
         switch (cmd) {
             case ALLOC_BUFFER: {
-                rpc_alloc_buffer(backend, input, output);
+                allocated_buffers.push_back(rpc_alloc_buffer(backend, input, output));
```
Comment (chraac) [inline ggml-rpc.cpp:941]: add the allocated buffer into list.

File: ggml-rpc.cpp [hunk 2]
```diff
@@ -950,7 +954,7 @@ static void rpc_serve_client(ggml_backend_t backend, sockfd_t sockfd, size_t fre
                 break;
             }
             case FREE_BUFFER: {
-                rpc_free_buffer(input);
+                allocated_buffers.remove(rpc_free_buffer(input));
```
Comment (chraac) [inline ggml-rpc.cpp:957]: remove the freed buffer from list

File: ggml-rpc.cpp [hunk 3]
```diff
@@ -993,6 +997,10 @@ static void rpc_serve_client(ggml_backend_t backend, sockfd_t sockfd, size_t fre
             break;
         }
     }
+
+    for (auto buff: allocated_buffers) {
+        ggml_backend_buffer_free(buff);
+    }
```
Comment (chraac) [inline ggml-rpc.cpp:1003]: free the reminding buffer.

**Comment by rgerganov** [comment]
I think a better approach would be to track allocated buffers with `std::unordered_set`, so we can also perform input validation `SET_TENSOR`, `GET_TENSOR, etc. I will submit a patch shortly and add you as a reviewer.

**Comment by chraac** [comment]
> I think a better approach would be to track allocated buffers with `std::unordered_set`, so we can also perform input validation in `SET_TENSOR`, `GET_TENSOR, etc. I will submit a patch shortly and add you as a reviewer.

emm, just forgot the `SET_TENSOR` and `GET_TENSOR` case, see there's also a tensor allocation in there.
also, for the `std::unordered_set`, thought the `std::list` will be okay here, since we won't have too many tensor, so it will no impact the free operation's performance.
agree that `std::unordered_set` is a better alternative here, let use it instead.

**Comment by chraac** [comment]
close this PR and wait for @rgerganov 's fix

**Pull request closed without merging**